### DISCUSSION TITLE
add type alias as a type member

### DISCUSSION THREAD

**Discussion by muhamad** [dotnet/csharplang#4247] [Language Ideas] [answered]
This is a proposal to add the `typedef` keyword to the language as a new member for class, struct and interface types, please see the next example

```cs
public interface IRecord<T>
{
    public T Id { get; set; }
}

public interface ITable<TRecord, TId> where TRecord : class, IRecord<TId>
{
    IEnumerable<TRecord> Records { get; }
}

public class Table<TRecord, TId> : ITable<TRecord, TId>
    where TRecord : class, IRecord<TId>
{
    private readonly IDictionary<TId, TRecord> _records;

    public IEnumerable<TRecord> Records { get; }
}
```

The above example define a table type and a record type, the record type define a generic argument which is the identity type.

When we define the `ITable` we specify that `TRecord` must implement `IRecord`, However the `IRecord` already have a generic argument which we must supply to `ITable` in order to use `IRecord` in the generic type constraint.

Now in order for `Table` to implement the `ITable` it must declare two generic arguments: the record type and its Id type, this design should be simplified because `Table` should have a mechanism to find the Id type of the record without generics at compile time, which leads to this proposal.

## The typedef solution

the `typedef` statement have following syntax:
- typedef &lt;type-name&gt; &lt;type-alias&gt; ;
- typedef &lt;type-alias&gt; ;

the first syntax is **type definition** which is to define an alias for a type and it is allowed in any struct, class and interface, the second syntax is to declare an alias which later must be defined and this syntax is allowed in interface and abstract classes.

The above example can now be written as the following:

```cs
public interface IRecord
{
    typedef IdType; // --> type alias declaration

    IdType Id { get; }
}

public interface ITable<TRecord> where TRecord : class, IRecord
{
    IEnumerable<TRecord> Records { get; }
}

public class Table<TRecord> : ITable<TRecord>
    where TRecord : class, IRecord
{
    private readonly IDictionary<TRecord.IdType, TRecord> _records;

    public IEnumerable<TRecord> Records { get; }
}

public class PersonRecord : IRecord
{
    typedef Guid IdType; // --> type alias definition

    public IdType Id { get; }
}
```

**Comment by huoyaoyuan** [comment, marked as answer]
This cannot be achieved under current CLR. Interface only contains vtable.
With CLR modification, this exact idea falls into the scope of type classes.

**Reply by 333fred** [reply]
Well, this is more of an associated type. It doesn't fall out of type classes necessarily, but we'll be thinking about this at the same time as type classes.
